### ISSUE TITLE
Rules with "too many matches" generate confusing "Timeout" error

### ISSUE THREAD

**Issue by jithurjacob** [semgrep/semgrep#2411]
**Describe the bug**
I am using semgrep in Azure Pipeline for PR analysis. Semgrep is getting timedout for a small JS file (~6K lines) in 5 seconds. I have seen multiple threads related to timeouts in JS but did not find any conclusion or plan of action. I have tried the command locally with only this file but no luck. Is there other workaround or any other option that I could use to get a scan no matter the time it takes? 

I have seen the use of -max_match_per_file  flag being used for semgrep-core but not part of the CLI. Is it possible to use this with CLI? Will this help with time outs?

**To Reproduce**
Unfortunately, I can not share the JS file. Here is the command. The failing rule is javascript.browser.security.raw-html-concat.raw-html-concat

semgrep   --config "p/clientside-js"  --timeout 0 --error --verbose --disable-version-check --include "********/application.js" --debug --strict

**Expected behaviour**
I expect the scan to complete as I've given timeout is 0.

**What is the priority of the bug to you?**
P0 - we are using semgrep to provide a report of secure coding practices to a client but since this single is not getting scanned we are unable to use semgrep.

**Environment**
Python package - 0.37.0

**Comment by aryx** [comment]
--timeout 0 means no timeout, which means the scan may not terminate at all because of a problematic file.
This is why we use a default of 5s for timeout, so that even one problematic file does not crash/take-forever the whole scan.

**Comment by jithurjacob** [comment]
@aryx - Yes but I think this should be left to the end-user, 5 seconds is too less. We should respect whatever value is provided by the end-user and give a warning saying it could take forever if the user provides 0. 5 seconds is arbitrary and there should be a way to override this one, maybe set another parameter timeout per file or exit after n matches.

I had given 3 hours as the time limit and later used 0 as I thought this would help. In my current setup, I'm running base scans(hardly 20 rules) in pre-commit and a large scan(600 rules) in PR scan. So even if the PR analysis takes 3 hours I'm happy.

**Comment by ievans** [comment]
@jithurjacob thanks for the detailed issue write-up! To me there are two issues:

- Timeout flag is not honored (as described)
- Semgrep takes more than 5s to run a rule on a single .js file

It sounds like the latter is something you can live with, but it is something that I'd like to make sure we fix as well. Is it at all possible for you to share a minimal/reduced/obfuscated version of the file that semgrep is slow on?

**Comment by jithurjacob** [comment]
@ievans - I'm sorry I won't be able to share the code. My organization strictly forbids it.

In some of the older issues related to timeout I had seen failed test cases against open source repos in which the scan is taking more than 5 seconds. Could you try with that?

**Comment by ievans** [comment]
@jithurjacob unfortunately I'm pretty sure we have fixed all of those already :/ @brendongo or @aryx is that true? If your organization is able to sign an NDA with r2c maybe that's a path forward, otherwise we'll just have to hope that increasing once the fix is in to increase timeout, it completes in finite time. 

@aryx maybe --profile in the core could be helpful here?

**Comment by jithurjacob** [comment]
@ievans - Unfortunately not possible, it's an enterprise and they are really slow. It would take atleast 6 months.

However, once the fix is done I can test and update you.

**Comment by aryx** [comment]
actually the default is 30s, not 5s.

**Comment by aryx** [comment]
But if you set --timeout 0, I don't see how you can get a timeout error on a JS file. Can you show the exact 
command line and the exact error message?
You should get a Parse error on the problematic file, not a timeout error.

**Comment by nbrahms** [comment]
Hi @jithurjacob ,

This should be fixed in Semgrep 0.39 and higher. In the meantime, you should be able to disable timeouts by setting `--timeout -1`.

**Comment by jithurjacob** [comment]
@aryx @nbrahms  - Screenshot for timeout error with both timeout 0 and -1. Just to clarify was -1 supposed to be in 0.39 or is it supposed to work in 0.38.0 itself? 

Please let me know if you need any additional information.

![image](https://user-images.githubusercontent.com/7101452/105364527-453a6e00-5c23-11eb-992a-da77e05a9e3e.png)

Edit: Added screenshot with time information

![image](https://user-images.githubusercontent.com/7101452/105365261-1375d700-5c24-11eb-8608-d81e5083ad93.png)

**Comment by nbrahms** [comment]
Hi @jithurjacob 

That's quite unexpected. I was able to reproduce your result using Semgrep 0.37.0 on `three.min.js`. It's interesting that this occurs only on a single rule. I'll investigate further.

For posterity:

```
✗ semgrep --config p/clientside-js --timeout -1 --lang js three.min.js
using config from https://semgrep.dev/p/clientside-js. Visit https://semgrep.dev/registry to see all public rules.
downloading config...
running 6 rules...
100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████|6/6
Warning: 1 timeout error(s) in three.min.js when running the following rules: [javascript.browser.security.raw-html-concat.raw-html-concat]
ran 6 rules on 1 files: 0 findings
1 files could not be analyzed; run with --verbose for details or run with --strict to exit non-zero if any file cannot be analyzed
```

**Comment by nbrahms** [comment]
It looks like we report "too many matches" as a timeout:
```
[0.083  Info       Main.Dune__exe__Main ] Version: semgrep-core version: v0.36.0-47-gd391669e, pfff: 0.42
[0.083  Info       Main.Dune__exe__Main ] Parsing /Users/nbrahms/tmp8ytlq9i0
[0.116  Info       Main.Dune__exe__Main ] processing 1 files
[0.116  Info       Main.Dune__exe__Main ] Analyzing three.js
[87.804 Info       Main.Dune__exe__Main ] done with three.js
[87.809 Info       Main.Dune__exe__Main ] found 128936 matches and 0 errors
[87.879 Info       Main.Dune__exe__Main ] too many matches on three.js, generating Timeout for it
[87.956 Info       Main.Dune__exe__Main ] most offending rule: id = 0..0.0.12.12.0.0.3, matches = 48370, pattern = function (...) {
  ...
  $X = (<... document ...>);
  ...
}
```

**Comment by nbrahms** [comment]
cc. @aryx I think we want to change the returned error to indicate that this won't be fixed by changing the timeout setting.

I propose `TooManyMatches` or some such name.

**Comment by nbrahms** [comment]
I reported this in the rule: https://github.com/returntocorp/semgrep-rules/issues/1074

**Comment by aryx** [comment]
yep, TooManyMatches sounds good.

**Comment by nbrahms** [comment]
@brendongo Is there a way to set the core's `-max_match_per_file` setting from the wrapper?

**Comment by nbrahms** [comment]
> yep, TooManyMatches sounds good.

Let's make sure the error message tells the user how to change the setting.

**Comment by brendongo** [comment]
> @brendongo Is there a way to set the core's `-max_match_per_file` setting from the wrapper?

Created https://github.com/returntocorp/semgrep/issues/2443 to track passing arbitrary opts to semgrep-core

**Comment by underyx** [comment]
> @brendongo Is there a way to set the core's `-max_match_per_file` setting from the wrapper?

This is blocking, FYI.

**Comment by underyx** [comment]
@aryx please also create an output confusion issue

**Comment by aryx** [comment]
This should work better now. People should get a TooManyMatches error, not a Timeout error anymore.

**Comment by jithurjacob** [comment]
@aryx @underyx @nbrahms  - Just to clarify, how will the user get issues related to a file in case of TooManyMatches error? 

Suppose a user scans three.js shown above with timeout 0, and semgrep got 128K matches. As per the new behaviour, semgrep will exit with TooManyMAtches and not Timeout. But how will this help end-users to identify actual issues? What should they do to get actual issues in case of too many matches let's say with a timeout of 0?

```
[0.083  Info       Main.Dune__exe__Main ] Version: semgrep-core version: v0.36.0-47-gd391669e, pfff: 0.42
[0.083  Info       Main.Dune__exe__Main ] Parsing /Users/nbrahms/tmp8ytlq9i0
[0.116  Info       Main.Dune__exe__Main ] processing 1 files
[0.116  Info       Main.Dune__exe__Main ] Analyzing three.js
[87.804 Info       Main.Dune__exe__Main ] done with three.js
[87.809 Info       Main.Dune__exe__Main ] found 128936 matches and 0 errors
[87.879 Info       Main.Dune__exe__Main ] too many matches on three.js, generating Timeout for it
[87.956 Info       Main.Dune__exe__Main ] most offending rule: id = 0..0.0.12.12.0.0.3, matches = 48370, pattern = function (...) {
  ...
  $X = (<... document ...>);
  ...
}```

**Comment by aryx** [comment]
The actual issue is that the pattern used is too general and match too many times, so we should fix this pattern/rule (see the "most offending rule" line in your example with the problematic pattern displayed).
Nathan actually recently improved some of our rules to use less of those general patterns while still finding the same final bugs.

**Comment by jithurjacob** [comment]
@aryx - Okay so if I understand correctly, semgrep found 128K matches due to a bad pattern in a rule. Since the number of patterns is greater than max_matches semgrep exits with TooManyMatches error. 

Suppose the user wants to find the issues no matter the wait time(overnight CI scan for example) and has defined timeout 0. How can they currently set max_matches to 0 or -1 to get the scan result?

**Comment by ievans** [comment]
@jithurjacob, I'm sure @aryx will answer your question but I thought I'd more about the philosophy behind Semgrep. We generally expect that eventually developers or security engineers are going to look at the output of the tool. 128K+ matches on a single file would take a few months to review if you are going to spend a minute on each finding! So I wouldn't expect us to optimize the performance of the engine to handle lots and lots of matches; we're much more concerned with improving the precision of rules so that fewer things are matched. 

But that's just how we expect the tool to be used. If you can share more about your use case I'm quite intrigued; is this feeding into another pipeline that will reduce the number of matches?

**Comment by jithurjacob** [comment]
@ievans I'm confused 128K matches means 128K issues have been found by a rule or 128K matches has been found by a rule which needs to be scanned by semgrep to find X issues?  I was under the impression that 128K matches need to be narrowed down by semgrep. So due to time constraints, TooManyMatches error was being thrown. 

No, we do not have a separate pipeline. It's a manual analysis of issues found by semgrep. We use about 600 rules in an Azure CI pipeline to check for potential issues in a PR. The codebase is about 60K lines of Python Flask and frontend. We got about 200 issues in rest of the code base of which only less than 10 needed to be ignored. 

We are actually stuck with one JS file which gets TimeoutError now and TooManyMatches in future due to [javascript.browser.security.raw-html-concat.raw-html-concat] rule. We need to provide a security audit report to our clients and getting this error doesn't look good. So my end goal is to give a successful scan report to clients even if it takes hours to scan I can remove the scan from PR to a weekly scan of the codebase.

**Comment by aryx** [comment]
The 128K is actually the number of match of a pattern in a rule. It does not mean indeed that semgrep at the end will return 128K matches, because this pattern is probably gonna be combined with other more restricting patterns in the rule.
I am actually working on reorganizing the engine to reorder how we look for those patterns which could solve the problem, but it's not yet ready.
In the mean time, we plan to have a way for the user to set the max_match_per_file via https://github.com/returntocorp/semgrep/issues/2443

**Comment by jithurjacob** [comment]
@aryx Thank you for the clarification. Is it possible for me to test this with SEMGREP_CORE_EXTRA? I haven't seen this in any documentation.

**Comment by aryx** [comment]
Yes. you can do export SEMGREP_CORE_EXTRA=-max_match_per_file 200000
The use of SEMGREP_CORE_PROFILE is documented in CONTRIBUTING.md but not SEMGREP_CORE_EXTRA indeed.
Those options are supposed to be used only by semgrep developers, but I guess in your case you can try it.

**Comment by jithurjacob** [comment]
@aryx - thanks this helped us along with setting the timeout as 0
